Compute C dependencies as a side-effect of compilation

dra27 opened the pull request.

dra27 committed:
It's all phony

Marking .depend as .PHONY reduces the amount of work does before
including them.

dra27 committed:
Remove C compilation from Makefile.common

dra27 committed:
Compute C dependencies as GCC intended

C dependencies are intended to be computed as a side-effect of compiling
a C file, not necessarily as a separate step. Compiling a C file with
GCC or clang is now responsible for directly creating the .d dependency
file, if dependencies are being computed. The workaround for MSVC now
becomes two calls to the C compiler, firstly to cl, to compile the file,
then to GCC to create the dependency information. This does the same
work as before, but with the calls in a different order.

Computing dependencies this way allows us to avoid re-checking the C
dependencies on every single invocation of make. If .d files are found,
then they are loaded. On first invocationm, neither the .o or the .d
file will exist, so make knows that the object needs rebuilding without
needing to compute the dependencies of it first.

dra27 committed:
Allow clang-cl to emit dependency information too

MisterDA commented on configure.ac:
```diff
@@ -623,14 +625,18 @@ LT_INIT
-      [DEP_CC],
-      [$DEP_CC gcc cc x86_64-w64-mingw32-gcc i686-w64-mingw32-gcc],
-      [false])],
+    [AS_IF([test x"$CC" = "xclang-cl"],
+      [DEP_CC="$CC"
+      DEP_CC_PREFIX='/clang:'],
```
```suggestion
    [AS_CASE(["$CC"],
      [clang-cl*],
        [DEP_CC="$CC"
        DEP_CC_PREFIX='/clang:'],
```
We could match on `clang-cl`, `clang-cl.exe` (especially if we're cross-compiling on a unix host, where there won't be a `.exe` extension). It's also possible that the executable is versioned with LLVM's version, as in `clang-cl-20` or `clang-cl-20.exe`. Maybe using the compiler vendor identifier is a better fit?

dra27 replied:
Oo - the vendor's a very good idea - you've made there be a concrete string in there for "clang-cl", right?

MisterDA replied:
`msvc-*-clang-*`

dra27 commented:
Rebased.

MisterDA commented on configure.ac:
```diff
@@ -724,14 +726,22 @@ popdef([host])dnl
+      DEP_CC_PREFIX='/clang:'],
+      [AC_CHECK_TOOLS(
+        [DEP_CC],
+        [$DEP_CC gcc cc x86_64-w64-mingw32-gcc i686-w64-mingw32-gcc],
```
```suggestion
        [$DEP_CC x86_64-w64-mingw32-gcc i686-w64-mingw32-gcc gcc cc],
```

I don't think this changes a lot but if we have includes guarded under `#ifdef _WIN32` they will be caught with mingw-w64, but not with Cygwin's gcc.

dra27 replied:
Hah - there is one place (in st_stubs.c) where we `#include` a local header guarded with `_WIN32`... but it's a good idea regardless!

dra27 replied:
Of course, perhaps the better hardening would simply be to add `-D_WIN32` as well when computing the deps?

MisterDA replied:
> Hah - there is one place (in st_stubs.c) where we `#include` a local header guarded with `_WIN32`... but it's a good idea regardless!

I did spend a few weeks digging in that file ;)

dra27 committed:
Check for triplet-prefixed compilers first

For the MSVC port, the triplet-prefixed versions of gcc will definitely
define _WIN32. Harden the use of, say, Cygwin gcc for dependencies by
ensuring _WIN32 is always defined.

MisterDA commented:
Passing by! you forgot to regenerate configure.
